### PR TITLE
Expanded configuration of quick-sip.

### PR DESCRIPTION
**Pull request by mandragorn**
all the cleanups in one place.  I've done my best to keep it backwards compatible, but I imagine I've screwed something up.  Where should I test the new version other than generator-torso-brec and my use-case in PSS?

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -28,16 +28,12 @@ After running the build task, a "dist" directory will be created:
...
+        |-- app.js
+        |-- app.scss
         |
         |-- index.html
 ```
-**app.js** is the browserified application that main.js completely specifies.
+**app.js** is the browserified application that app.js completely specifies.
 **core.css & whatever.css** are the sass-lib compiled css files.
 
 ## Executing Tasks
```
Comment (mandragorn) [inline README.md:42]: I need to make more changes to this file.  But the rest is ready for review.

File: index.js [hunk 2]
```diff
@@ -1,83 +1,144 @@
...
+ *    @param [options.browserify.root='./<options.src>/scripts/main'] {String} the root file to bundle.
+ *    @param [options.browserify.out='app.js'] {String} name for the bundled file (or full path of the bundled file).
+ *    @param [options.browserify.failOnError=false] {Boolean} whether the browserify pipe should fail when browserifying fails.
+ *    @param [options.browserify.debug=$.util.env.type !== 'production'] {Boolean} whether to run browserify in debug mode.
+ *                                                                       Defaults to true if the env type is production, false otherwise.
+ *    @param [options.browserify.dist='<options.dist>/scripts] {String} name for the bundled file (or full path of the bundled file).
+ *  @param [options.styles] {Object} configuration options for the styles (build-styles) task.
+ *    @param [options.styles.skip=false] {Boolean} whether to skip (true) or run (false) the styles task.
+ *    @param [options.styles.src='<options.src>/**//*.scss'] {String} all styles source files to include.
+ *    @param [options.styles.root='<options.src>/app'] {String} the root styles file to bundle.
```
Comment (kentmw) [inline index.js]: The README needs to break down the options by task in a clear way

File: index.js [hunk 3]
```diff
@@ -1,83 +1,144 @@
...
+    taskPrefix: '',
+    src: 'app',
+    dist: 'dist',
+    clean: {}, // see defaults in ./tasks/clean.js - only included here for the skips below.
+    browserify: {}, // see defaults in ./tasks/build-app.js - only included here for the skips below.
+    styles: {}, // see defaults in ./tasks/build-styles.js - only included here for the skips below.
+    copy: {} // see defaults in ./tasks/copy-resources.js - only included here for the skips below.
+  }, options);
+
+  var derivedDefaultCleanOptions = {
```
Comment (kentmw) [inline index.js]: not a fan of the name "derivedDefault" can we call it bundleDefault? or maybe:

```
  var bundleDefaults: {
    clean: {

    },
   browserify: {

   },
   ...
}
```

File: index.js [hunk 4]
```diff
@@ -1,83 +1,144 @@
...
-    return gulp.src(paths.dist, {read: false})
-      .pipe($.rimraf());
+    if (options.clean.skip) {
+      runSequence(buildTasks, browserifyCompleteFn);
+    } else {
+      runSequence(options.taskPrefix + 'clean', buildTasks, browserifyCompleteFn);
+    }
   });
+  // Alias default to do the build.  After this file is run the default task can be overridden if desired.
+  gulp.task('default', ['build']);
```
Comment (kentmw) [inline index.js]: This is where I expect this file to end. Should the watch task be in it's own file?

File: index.js [hunk 5]
```diff
@@ -96,101 +157,66 @@ module.exports = function(gulp) {
...
-
-    nonResources: function(nrsc) {
-      nonResources = nrsc;
+    /**
+     * @method transform
+     * @param transform {String or Object} either a string referencing the transform to perform or an object
+     *                  can either be a string identifying the transform or an object of the format:
+     *                    {name: 'name of transform', options: {// transform options}}
+     */
+    transform: function(transform) {
```
Comment (kentmw) [inline index.js]: now that you've added ability to set options in the constructor, is it necessary to have any of these public API calls?

File: tasks/build-styles.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
+/**
+ * @class CopyResourcesTask
+ * @constructor
+ *  @param [options] {Object} configuration options for the styles (build-styles) task.
+ *    @param [options.skip=false] {Boolean} whether to skip (true) or run (false) the styles task.
+ *    @param [options.src='<options.src>/**//*.scss'] {String} all styles source files to include.
```
Comment (kentmw) [inline tasks/build-styles.js]: README will be important to explain how it looks like to use these tasks

**Comment by kentmw** [comment]
@mandragorn , I like where we are going with this. Some naming issues and README requirements, but otherwise nice. See comments above

File: index.js (hunk 5 above)
Comment (mandragorn) [inline index.js]: I was keeping backwards compatibility.

File: index.js [hunk 7]
```diff
@@ -1,196 +1,64 @@
...
+    nonResources: function(excludes) {
+      bundleOptions = mergeBundleDefaults({
+        browserify: {
+          excludes: excludes
+        }
+      });
 
-    configurePaths: function(options) {
-      _.extend(paths, options);
+      buildTask.options(bundleOptions);
```
Comment (mandragorn) [inline index.js]: I need to change this api to a pure callthrough to buildTask.nonResources() and watchTask.nonResources(), same with the rest of the api (options() and transform()).

File: index.js (hunk 5 above)
Comment (mandragorn) [inline index.js]: Do we want to preserve backwards compatibility or can we drop that?  I keep running into corner cases when trying to support changing the configuration after the tasks are created.

**Comment by mandragorn** [comment]
This would be much easier if we just drop the public api calls and just have the generation of tasks based on the options passed in.  If you want to change the options, you just initialize quick-sip again.

@kentmw @arikwex what are your thoughts?

**Comment by mandragorn** [comment]
At this point this is ready to be merged.  Let me know if you have any comments on it.

@arikwex @kentmw

**Comment by mandragorn** [comment]
And you can actually read the readme here: https://github.com/mandragorn/quick-sip

**Pull request merged**